chainstate/src/detail/bootstrap.rs

```diff
@@ -0,0 +1,143 @@
+    #[error("Block import error: {0}")]
+    BlockProcessing(#[from] BlockError),
+    #[error("Block import error: {0}")]
+    FailedToReadProperty(#[from] PropertyQueryError),
+}
+
+impl From<std::io::Error> for BootstrapError {
+    fn from(error: std::io::Error) -> Self {
+        Self::File(error.to_string())
+    }
+}
+
+pub fn import_bootstrap_stream<P, S: std::io::Read>(
+    expected_magic_bytes: &[u8],
+    file_reader: &mut std::io::BufReader<S>,
+    process_block_func: &mut P,
+    chainstate_config: &ChainstateConfig,
+) -> Result<(), BootstrapError>
+where
+    P: FnMut(WithId<Block>) -> Result<Option<BlockIndex>, BlockError>,
+{
+    // min: The smallest buffer size, after which another read is triggered from the bootstrap file
+    // max: The largest buffer size, after which reading the file is stopped
+    // NOTE: both sizes MUST be larger than the largest block in the blockchain + 4 bytes for magic bytes
+    let (min_buffer_size, max_buffer_size) =
+        chainstate_config.min_max_bootstrap_import_buffer_sizes.unwrap_or((
+            DEFAULT_MIN_IMPORT_BUFFER_SIZE,
+            DEFAULT_MAX_IMPORT_BUFFER_SIZE,
+        ));
+
+    // It's more reasonable to use a VeqDeque, but it's incompatible with the windows() method which is needed to search for magic bytes
+    // There's a performance hit behind this, but we don't care. Anyone is free to optimize this.
+    let mut buffer_queue = Vec::<u8>::new();
+
+    loop {
+        if buffer_queue.len() < min_buffer_size {
+            fill_buffer(&mut buffer_queue, file_reader, max_buffer_size)?;
+        }
+
+        // locate magic bytes to recognize the start of a block
+        let current_pos = buffer_queue
+            .windows(expected_magic_bytes.len())
+            .position(|window| window == expected_magic_bytes);
+
+        // read the block after the magic bytes
+        let block = match current_pos {
+            Some(v) => read_block_at_pos(&buffer_queue[v + expected_magic_bytes.len()..])?,
+            None => break,
+        };
+        let block_len = block.encoded_size();
+        process_block_func(block.into())?;
+
+        // consume the buffer from the front
+        buffer_queue = buffer_queue[expected_magic_bytes.len() + block_len..].to_vec();
+    }
+
+    Ok(())
+}
+
+fn fill_buffer<S: std::io::Read>(
+    buffer_queue: &mut Vec<u8>,
+    reader: &mut std::io::BufReader<S>,
+    max_buffer_size: usize,
+) -> Result<(), BootstrapError> {
+    while buffer_queue.len() < max_buffer_size {
+        let buf_len = {
+            let data = reader.fill_buf()?;
+            if data.is_empty() {
+                break;
+            }
+            buffer_queue.extend(data.iter());
+            data.len()
+        };
+        reader.consume(buf_len);
+    }
+
+    Ok(())
+}
+
+fn read_block_at_pos(buf: &[u8]) -> Result<Block, BootstrapError> {
+    let mut buffer = buf;
+    let block = Block::decode(&mut buffer)?;
```

Comment:
I understand that this is from 3rd-party code, but I'm curious why `decode` should require mutability?

Reply:
The slice itself is immutable, but we a copy of the slice to track how the decoding progresses. Imagine the function parameter to be a `const uint8_t*` with a size associated with it (something like `std::span` or `std::string_view`), and the `buffer` in line 116 to be a `uint8_t*` starting at the same starting point of `buf`. While you're decoding, you need to increment some pointer, and the `buffer` pointer is the one you increment. In rust, the second pointer is also a slice that keeps changing and reducing its size while the decoding is happening. So you have the original slice `buf`, and another shrinking slice `buffer` that tracks the position to consume the bytes while decoding.

Reply:
oh, right it's just a pointer that is copied not the data... thanks